generators/E906LegacyGen/SQPileupGen.C

```diff
@@ -0,0 +1,195 @@
+_evtID(evtID),
+_pdg(pdg),
+  _pos(pos),
+  _mom(mom)
+{
+}
+
+SQPileupGen::SQPileupGen(const std::string& name):
+  PHG4ParticleGeneratorBase(name),
+  _ineve(nullptr),
+  _mcevt(nullptr),
+  _extFile(nullptr),
+  _extTree(nullptr),
+  _extPos(nullptr),
+  _extMom(nullptr),
+  _bucketSize(0),
+  _evt(0),
+  _beam_intensity_profile(nullptr),
+  _beam_intensity_profile_histo(nullptr),
+  _n_proc_evt(0),
+  _inhibit_threshold(5000),
+  _proton_coeff(0.03),
+  _vertexGen(new SQPrimaryVertexGen)
+{
+}
+
+SQPileupGen::~SQPileupGen()
+{
+  if(_vertexGen != nullptr) delete _vertexGen;
+}
+
+int SQPileupGen::Init(PHCompositeNode* topNode)
+{  
+  return Fun4AllReturnCodes::EVENT_OK;
+}
+
+int SQPileupGen::InitRun(PHCompositeNode* topNode)
+{
+
+  gRandom->SetSeed(PHRandomSeed());
+  _ineve = findNode::getClass<PHG4InEvent>(topNode, "PHG4INEVENT");
+  if(!_ineve) 
+    {
+      PHNodeIterator iter(topNode);
+      PHCompositeNode* dstNode = dynamic_cast<PHCompositeNode*>(iter.findFirst("PHCompositeNode", "DST"));
+      
+      _ineve = new PHG4InEvent();
+      dstNode->addNode(new PHDataNode<PHObject>(_ineve, "PHG4INEVENT", "PHObject"));
+    }
+
+  _evt = findNode::getClass<SQEvent>(topNode, "SQEvent");
+  if (! _evt) {
+    PHNodeIterator iter(topNode);
+    PHCompositeNode* dstNode = dynamic_cast<PHCompositeNode*>(iter.findFirst("PHCompositeNode", "DST"));
+
+    _evt = new SQEvent_v1();
+    dstNode->addNode(new PHIODataNode<PHObject>(_evt, "SQEvent", "PHObject"));
+  }
+
+  _mcevt = findNode::getClass<SQMCEvent>(topNode, "SQMCEvent");
+  if(!_mcevt) 
+    {
+      PHNodeIterator iter(topNode);
+      PHCompositeNode* dstNode = dynamic_cast<PHCompositeNode*>(iter.findFirst("PHCompositeNode", "DST"));
+
+      _mcevt = new SQMCEvent_v1();
+      dstNode->addNode(new PHIODataNode<PHObject>(_mcevt, "SQMCEvent", "PHObject"));
+    }
+
+  _extFile = TFile::Open(_extFileName.Data());
+  if(!_extFile)
+    {
+      std::cout << "SQPileupGen: cannot find the input ROOT file, abort " << _extFileName << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  _extTree = (TTree*)_extFile->Get("bkg");
+
+  _extPos = new TClonesArray("TVector3");
+  _extMom = new TClonesArray("TVector3");
+
+  _extTree->SetBranchAddress("eventID", &_extEvtID);
+  _extTree->SetBranchAddress("n", &_nExtPar);
+  _extTree->SetBranchAddress("pdg", _extPDG);
+  _extTree->SetBranchAddress("pos", &_extPos);
+  _extTree->SetBranchAddress("mom", &_extMom);
+
+  _readIdx = 0;
+
+  _vertexGen->InitRun(topNode);
+  return Fun4AllReturnCodes::EVENT_OK;
+}
+
+int SQPileupGen::process_event(PHCompositeNode* topNode)
+{
+  if(_beam_intensity_profile) 
+    {
+      setBucketSize((int)_beam_intensity_profile->GetRandom()/_proton_coeff);//if the function describes dN/dI vs I   
+    }
+
+  if(_beam_intensity_profile_histo)
+    {
+      double QIEcount = _beam_intensity_profile_histo->GetRandom();
+      if (QIEcount>_inhibit_threshold) return Fun4AllReturnCodes::ABORTEVENT;
```

Comment:
Is it possible to apply this cut in case of `_beam_intensity_profile`?  It should be helpful when we compare profiles.

Is there a good reason of using `return` here, rather than repeating `GetRandom()` until `QIEcount < _inhibit_threshold`?

Reply:
I was wondering if you mean to apply in the histogram itself.  Could you please make both of the comments more clear. Thank you.

Reply:
I mean to apply the cut in case of TF1, when  `_beam_intensity_profile` is true.

The 2nd comment means the following change for example.  It could be improper in terms of normalization.
```
double QIEcount;
do {
  QIEcount = _beam_intensity_profile_histo->GetRandom();
} while (QIEcount>_inhibit_threshold);
```

Reply:
1st comment: I see. It should be possible. `void TF1::SetRange(Double_t  xmin, Double_t  xmax)` with xmax being our threshold cut. 


2nd comment: It would be a loop within an event. I was wondering why we would want that loop within an event. I believe, we want to have a distribution of the QIE count over the run not over the event. Don't we?